healpix_alchemy/healpix.py

```diff
@@ -0,0 +1,116 @@
+"""Model base classes for multiresolution HEALPix data."""
+from astropy.coordinates import ICRS
+from astropy import units as u
+from astropy_healpix import (level_to_nside,
+                             uniq_to_level_ipix, HEALPix)
+from mocpy import MOC
+from sqlalchemy import BigInteger, Column
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.ext.hybrid import hybrid_property
+from sqlalchemy.orm import relationship
+from sqlalchemy.orm.mapper import Mapper
+
+LEVEL = MOC.HPY_MAX_NORDER
+"""Base HEALPix resolution. This is the maximum HEALPix level that can be
+stored in a signed 8-byte integer data type."""
+
+HPX = HEALPix(nside=level_to_nside(LEVEL), order='nested', frame=ICRS())
+"""HEALPix projection object."""
+
+PIXEL_AREA = HPX.pixel_area.to_value(u.sr)
+"""Native pixel area in steradians."""
+
+
+class Tile:
+    """Mixin class for a table that stores a HEALPix multiresolution tile."""
+
+    nested_lo = Column(
+        BigInteger, index=True, nullable=False, primary_key=True,
+        doc=('Lower end of range of HEALPix nested indices at',
+             f'nside=2**{LEVEL}'))
+
+    nested_hi = Column(
+        BigInteger, nullable=False,
+        doc=('Upper end of range of HEALPix nested indices at',
+             f'nside=2**{LEVEL}'))
+
+    def __init__(self, *args, uniq=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        if uniq is not None:
+            self.uniq = uniq
+
+    @hybrid_property
+    def uniq(self):
+        """HEALPix UNIQ pixel index."""
+        # This is the same expression as in astropy_healpix.level_ipix_to_uniq,
+        # but reproduced here so that SQLAlchemy can map it to SQL.
+        ipix = self.nested_lo
+        # FIXME: wrong, get level from log2(nested_hi - nested_lo)/2
+        return int(ipix + (1 << 2 * (LEVEL + 1)))
```

Comment:
Indeed, this is wrong.

Reply:
Ok updated with the expression in that line.

Reply:
Still wrong... this script:

```python
#!/usr/bin/env python
from astropy_healpix import uniq_to_level_ipix, level_ipix_to_uniq
from mocpy import MOC
import numpy as np

LEVEL = MOC.HPY_MAX_NORDER

value = level_ipix_to_uniq(3, 12)
print('expected', value)

# @uniq.setter
level, ipix = uniq_to_level_ipix(value)
shift = 2 * (LEVEL - level)
nested_lo = int(ipix << shift)
nested_hi = int(((ipix + 1) << shift) - 1)

# @uniq
ipix = nested_lo
level = int(np.log2(nested_hi - nested_lo)/2)
uniq = int(ipix + (1 << 2 * (level + 1)))

print('got', uniq)
```

produces this output:

```
expected 268
got 72057594037927936
```

 The expression has to involve both `level` and `LEVEL`, like in the setter.

Reply:
How about something like the below?

`LEVEL = MOC.HPY_MAX_NORDER

value = level_ipix_to_uniq(3, 12)
print('expected', value)

# @uniq.setter
level, ipix = uniq_to_level_ipix(value)
shift = 2 * (LEVEL - level)
nested_lo = int(ipix << shift)
nested_hi = int(((ipix + 1) << shift) - 1)

# @uniq
level = LEVEL - (int(np.ceil(np.log2(nested_hi - nested_lo)/2)))
shift = 2 * (LEVEL - level)
ipix = int(nested_lo >> shift)
uniq = int(ipix + (1 << 2 * (level + 1)))
print('got', uniq)`

Reply:
Looks plausible. How about a unit test?